File: birzeit/palgrid.js
/* تحويل الإحداثيات: WGS84 ⇄ فلسطين 1923 / الشبكة الفلسطينية (EPSG:28191).
 * نفس معادلات ومعاملات tools/palgrid.py (اللي بنى ملفات البيانات)، حتى الأرقام المعروضة تطابق الأصل.
 * Cassini-Soldner على Clarke 1880 (Benoit) + Helmert بسبع معاملات (towgs84 القياسي). */
(function () {
  "use strict";
  var A = 6378300.789, B = 6356566.435, E2 = 1 - (B * B) / (A * A);
  var LAT0 = 31.73409694444445 * Math.PI / 180, LON0 = 35.21208055555556 * Math.PI / 180;
  var FE = 170251.555, FN = 126867.909;
  var TX = -275.7224, TY = 94.7824, TZ = 340.8944, S = 1e-6;
  var SEC = Math.PI / 180 / 3600, RX = -8.001 * SEC, RY = -4.42 * SEC, RZ = -11.821 * SEC;
  var WA = 6378137.0, WF = 1 / 298.257223563, WE2 = WF * (2 - WF);

  function mer(phi) {
    var e2 = E2, e4 = e2 * e2, e6 = e4 * e2;
    return A * ((1 - e2 / 4 - 3 * e4 / 64 - 5 * e6 / 256) * phi
      - (3 * e2 / 8 + 3 * e4 / 32 + 45 * e6 / 1024) * Math.sin(2 * phi)
      + (15 * e4 / 256 + 45 * e6 / 1024) * Math.sin(4 * phi)
      - (35 * e6 / 3072) * Math.sin(6 * phi));
  }
  var M0 = mer(LAT0);

  function toEcef(phi, lam, a, e2) {
    var s = Math.sin(phi), nu = a / Math.sqrt(1 - e2 * s * s);
    return [nu * Math.cos(phi) * Math.cos(lam), nu * Math.cos(phi) * Math.sin(lam), nu * (1 - e2) * s];
  }
  function fromEcef(p, a, e2) {
    var lam = Math.atan2(p[1], p[0]), r = Math.hypot(p[0], p[1]), phi = Math.atan2(p[2], r * (1 - e2));
    for (var i = 0; i < 6; i++) {
      var s = Math.sin(phi), nu = a / Math.sqrt(1 - e2 * s * s);
      phi = Math.atan2(p[2] + e2 * nu * s, r);
    }
    return [phi, lam];
  }
  // WGS84 → فلسطين 1923 (عكس towgs84)
  function helmertInv(p) {
    var x = p[0] - TX, y = p[1] - TY, z = p[2] - TZ, sc = 1 - S;
    return [sc * (x + RZ * y - RY * z), sc * (-RZ * x + y + RX * z), sc * (RY * x - RX * y + z)];
  }

  /** (lon, lat) WGS84 → [E, N] بالمتر على الشبكة الفلسطينية */
  function fromWgs84(lon, lat) {
    var g = fromEcef(helmertInv(toEcef(lat * Math.PI / 180, lon * Math.PI / 180, WA, WE2)), A, E2);
    var phi = g[0], lam = g[1];
    var s = Math.sin(phi), nu = A / Math.sqrt(1 - E2 * s * s), t = Math.pow(Math.tan(phi), 2);
    var c = E2 * Math.pow(Math.cos(phi), 2) / (1 - E2), a = (lam - LON0) * Math.cos(phi);
    var x = nu * (a - t * Math.pow(a, 3) / 6 - (8 - t + 8 * c) * t * Math.pow(a, 5) / 120);
    var y = mer(phi) - M0 + nu * Math.tan(phi) * (a * a / 2 + (5 - t + 6 * c) * Math.pow(a, 4) / 24);
    return [FE + x, FN + y];
  }

  window.PalGrid = { fromWgs84: fromWgs84 };
})();

File: birzeit/tools/palgrid.py
"""تحويل الإحداثيات: فلسطين 1923 / الشبكة الفلسطينية (EPSG:28191) ⇄ WGS84.

المعاملات القياسية (EPSG):
- الإسقاط: Cassini-Soldner، مركزه القدس، على إهليلج Clarke 1880 (Benoit).
- تحويل المرجع: Helmert بسبع معاملات (towgs84 كما في proj لـ EPSG:28191، اتفاقية Position Vector).
نفس المعادلات منقولة للواجهة (birzeit/app/palgrid.js) حتى تطابق الأرقام بالموقع ما يولّده هذا الملف.
"""
import math

A = 6378300.789
B = 6356566.435
E2 = 1 - (B * B) / (A * A)
LAT0 = math.radians(31.73409694444445)
LON0 = math.radians(35.21208055555556)
FE = 170251.555
FN = 126867.909
# towgs84 (متر، ثواني قوسية، جزء بالمليون)
TX, TY, TZ = -275.7224, 94.7824, 340.8944
RX, RY, RZ = (math.radians(v / 3600) for v in (-8.001, -4.42, -11.821))
S = 1e-6

WA = 6378137.0
WF = 1 / 298.257223563
WE2 = WF * (2 - WF)


def _m(phi):
    e2, e4, e6 = E2, E2 * E2, E2 ** 3
    return A * ((1 - e2 / 4 - 3 * e4 / 64 - 5 * e6 / 256) * phi
                - (3 * e2 / 8 + 3 * e4 / 32 + 45 * e6 / 1024) * math.sin(2 * phi)
                + (15 * e4 / 256 + 45 * e6 / 1024) * math.sin(4 * phi)
                - (35 * e6 / 3072) * math.sin(6 * phi))


M0 = _m(LAT0)


def cassini_inverse(e, n):
    m1 = M0 + (n - FN)
    mu = m1 / (A * (1 - E2 / 4 - 3 * E2 ** 2 / 64 - 5 * E2 ** 3 / 256))
    e1 = (1 - math.sqrt(1 - E2)) / (1 + math.sqrt(1 - E2))
    phi1 = (mu + (3 * e1 / 2 - 27 * e1 ** 3 / 32) * math.sin(2 * mu)
            + (21 * e1 ** 2 / 16 - 55 * e1 ** 4 / 32) * math.sin(4 * mu)
            + (151 * e1 ** 3 / 96) * math.sin(6 * mu)
            + (1097 * e1 ** 4 / 512) * math.sin(8 * mu))
    s = math.sin(phi1)
    nu1 = A / math.sqrt(1 - E2 * s * s)
    rho1 = A * (1 - E2) / (1 - E2 * s * s) ** 1.5
    t1 = math.tan(phi1) ** 2
    d = (e - FE) / nu1
    phi = phi1 - (nu1 * math.tan(phi1) / rho1) * (d * d / 2 - (1 + 3 * t1) * d ** 4 / 24)
    lam = LON0 + (d - t1 * d ** 3 / 3 + (1 + 3 * t1) * t1 * d ** 5 / 15) / math.cos(phi1)
    return phi, lam


def cassini_forward(phi, lam):
    s = math.sin(phi)
    nu = A / math.sqrt(1 - E2 * s * s)
    t = math.tan(phi) ** 2
    c = E2 * math.cos(phi) ** 2 / (1 - E2)
    a = (lam - LON0) * math.cos(phi)
    x = nu * (a - t * a ** 3 / 6 - (8 - t + 8 * c) * t * a ** 5 / 120)
    y = _m(phi) - M0 + nu * math.tan(phi) * (a * a / 2 + (5 - t + 6 * c) * a ** 4 / 24)
    return FE + x, FN + y


def _to_ecef(phi, lam, a, e2):
    s = math.sin(phi)
    nu = a / math.sqrt(1 - e2 * s * s)
    return nu * math.cos(phi) * math.cos(lam), nu * math.cos(phi) * math.sin(lam), nu * (1 - e2) * s


def _from_ecef(x, y, z, a, e2):
    lam = math.atan2(y, x)
    p = math.hypot(x, y)
    phi = math.atan2(z, p * (1 - e2))
    for _ in range(6):
        s = math.sin(phi)
        nu = a / math.sqrt(1 - e2 * s * s)
        phi = math.atan2(z + e2 * nu * s, p)
    return phi, lam


def _helmert(x, y, z, sign):
    rx, ry, rz, sc = sign * RX, sign * RY, sign * RZ, 1 + sign * S
    if sign > 0:
        return (TX + sc * (x - rz * y + ry * z),
                TY + sc * (rz * x + y - rx * z),
                TZ + sc * (-ry * x + rx * y + z))
    x, y, z = x - TX, y - TY, z - TZ
    return (sc * (x - rz * y + ry * z), sc * (rz * x + y - rx * z), sc * (-ry * x + rx * y + z))


def to_wgs84(e, n):
    """(E, N) بالشبكة الفلسطينية ← (lon, lat) WGS84 بالدرجات."""
    phi, lam = cassini_inverse(e, n)
    x, y, z = _helmert(*_to_ecef(phi, lam, A, E2), 1)
    phi, lam = _from_ecef(x, y, z, WA, WE2)
    return math.degrees(lam), math.degrees(phi)


def from_wgs84(lon, lat):
    """(lon, lat) WGS84 ← (E, N) بالشبكة الفلسطينية."""
    x, y, z = _helmert(*_to_ecef(math.radians(lat), math.radians(lon), WA, WE2), -1)
    phi, lam = _from_ecef(x, y, z, A, E2)
    return cassini_forward(phi, lam)


if __name__ == '__main__':
    for e, n in [(FE, FN), (168700, 153400)]:
        lon, lat = to_wgs84(e, n)
        e2, n2 = from_wgs84(lon, lat)
        print(f'{e:.3f},{n:.3f} -> {lat:.7f},{lon:.7f} -> back {e2 - e:.4f} {n2 - n:.4f} m')
